docs/metrics.md

```diff
@@ -0,0 +1,561 @@
+rather than creating vectors. Arrays are more efficient as they avoid
+unnecessary heap allocations on the measurement path:
+
+```rust
+// Good practice: Using an array slice directly
+counter.add(2, &[KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+
+// Avoid this: Creating a Vec is unnecessary, and it allocates on the heap
+// counter.add(2, &vec![KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+```
+
+This approach minimizes memory allocations on metrics hot paths, improving
+overall performance when reporting metrics frequently.
+
+### Reporting measurements via asynchronous instruments
+
+Asynchronous instruments like `ObservableCounter` are ideal for reporting
+metrics that are already being tracked or stored elsewhere in your application.
+These instruments allow you to observe and report the current state of existing
+values without manually managing the aggregation yourself.
+
+:heavy_check_mark: Use `ObservableCounter` when you already have a variable
+tracking a count:
+
+```rust
+// Example: Using ObservableCounter when you already have a variable tracking counts
+use opentelemetry::KeyValue;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+// An existing variable in your application
+static REQUEST_COUNT: AtomicU64 = AtomicU64::new(0);
+
+// In your application code, you update this directly
+fn handle_request() {
+    // Process request...
+    REQUEST_COUNT.fetch_add(1, Ordering::SeqCst);
+}
+
+// When setting up metrics, register an observable counter that reads from your variable
+fn setup_metrics(meter: &opentelemetry::metrics::Meter) {
+    let _observable_counter = meter
+        .u64_observable_counter("request_count")
+        .with_description("Number of requests processed")
+        .with_unit("requests")
+        .with_callback(|observer| {
+            // Read the current value from your existing counter
+            observer.observe(
+                REQUEST_COUNT.load(Ordering::SeqCst), 
+                &[KeyValue::new("endpoint", "/api")]
+            )
+        })
+        .build();
+}
+```
+
+:heavy_check_mark: Use regular Counter when you need to increment counts at
+specific points in your code:
+
+```rust
+// Example: Using Counter when incrementing at specific code points
+use opentelemetry::KeyValue;
+
+fn setup_metrics(meter: &opentelemetry::metrics::Meter) -> opentelemetry::metrics::Counter<u64> {
+    meter
+        .u64_counter("processed_items")
+        .with_description("Number of items processed")
+        .build()
+}
+
+// Then in your processing code:
+fn process_item(counter: &opentelemetry::metrics::Counter<u64>, item_type: &str) {
+    // Process item...
+    
+    // Increment the counter with the item type as an attribute
+    counter.add(1, &[KeyValue::new("type", item_type)]);
+}
+```
+
+## MeterProvider Management
+
+:stop_sign: Avoid creating `MeterProvider` instances too frequently. A
```

Comment:
Similar to this: https://github.com/open-telemetry/opentelemetry-rust/pull/2946#discussion_r2067407462

I think we should start with something like: "Most use-cases require you to create ONLY one instance of `MeterProvider`. You should NOT create multiple instances of `MeterProvider` unless you have some unusual requirement of having different export strategies within the same application. Using multiple instances of `MeterProvider` requires users to exercise caution."